### PR TITLE
Refactor constraint violations

### PR DESCRIPTION
**Pull request by benjamin-shen**
### Summary 

This pull request implements constraint violations in a more generalized way.

- [x] refactored constraint violations from `removeIllegalEdgesFromRequirementFulfillmentGraph()`
  - added `requirementsThatDoNotAllowDoubleCounting` from the constraint violation computation to the vuex store
  - the constraint violation computation does not rely on the dangerous and safe graphs, it can take arbitrary graphs (as is needed to fix the requirements bug in #626)
- [x] modified `allowDoubleCounting` function to account for major/minor/grad
  - this works by reframing double counting as a relation (can a course double count for requirement A and requirement B?) rather than a feature of a single requirement (does requirement A allow double counting with all other requirements?)
  - there are some places we need to take a closer look at the usage of `.allowCourseDoubleCounting`, because it's most likely incorrect

Remaining TODOs:

- [ ] fix #626's requirements bug

### Test Plan 

No major frontend changes, hopefully.

### CODE REVIEW

**Review by willespencer — CHANGES_REQUESTED** [review]
Left a bunch of comments! Your explanation really helped get through this and very excited for these changes 😄 

Also tested some things detailed below

1. I checked how double counting affects minors (specifically the Human Development minor).

A big improvement is how the double counting issue is fixed for minors - so now since the two requirements do not have the double counting flag, courses from the add modal must select only one - unlike on CoursePlan where one course can fulfill both reqs.
image

However, when added from the add modal, for some reason this course cannot be applied to both `Liberal Studies` and the minor unlike prod even though this should be allowed I believe. Can this be fixed before the new add modal?
image

2. Requirements/Double counting seems to work as expected in other cases! For example, `CS 1110` fulfills the computing requirement in Eng and CS simultaneously, `CS 4850` fulfills two reqs in CS, and a class like `ECON 3040` can be applied to MPA or Engineering, but not both.

File: src/requirements/requirement-frontend-computation.ts [hunk 1]
```diff
@@ -238,6 +238,7 @@ export default function computeGroupedRequirementFulfillmentReports(
   readonly userRequirementsMap: Readonly<Record<string, RequirementWithIDSourceType>>;
   readonly dangerousRequirementFulfillmentGraph: RequirementFulfillmentGraph<string, CourseTaken>;
   readonly safeRequirementFulfillmentGraph: RequirementFulfillmentGraph<string, CourseTaken>;
+  readonly requirementsThatDoNotAllowDoubleCounting: Map<string | number, Set<string[]>>;
```
[the remark this answers is not in the document]
Comment (benjamin-shen) [reply src/requirements/requirement-frontend-computation.ts:241]: The mapping is from course unique id (`string | number`) to requirement id (`string`).

>mapping a requirement to a set of requirements that cannot be double counted with it

This is an intermediary step in `getConstraintViolationsForSingleCourse()` to generate `requirementsThatDoNotAllowDoubleCounting`, the variable name is `constraintViolatingRequirements`.

Should I rename the variable(s)?

File: src/requirements/requirement-frontend-computation.ts [hunk 2]
```diff
@@ -357,9 +360,10 @@ export default function computeGroupedRequirementFulfillmentReports(
   }
 
   return {
-    userRequirementsMap: Object.fromEntries(userRequirements.map(it => [it.id, it])),
+    userRequirementsMap,
```
[the remark this answers is not in the document]
Comment (benjamin-shen) [reply src/requirements/requirement-frontend-computation.ts:363]: `userRequirementsMap` is computed the same way and returned in `buildRequirementFulfillmentGraphFromUserData()`, so I'm just removing repeated code.

File: src/requirements/requirement-frontend-utils.ts [hunk 3]
```diff
@@ -86,6 +89,37 @@ export function requirementAllowDoubleCounting(
   return requirement.allowCourseDoubleCounting || false;
 }
 
+export function allowCourseDoubleCountingBetweenRequirements(
+  requirementA: RequirementWithIDSourceType,
+  requirementB: RequirementWithIDSourceType
+): boolean {
+  const allowCourseDoubleCounting =
+    requirementA.allowCourseDoubleCounting || requirementB.allowCourseDoubleCounting || false;
```
[the remark this answers is not in the document]
Comment (benjamin-shen) [reply src/requirements/requirement-frontend-utils.ts:97]: The `allowCourseDoubleCounting` property can be undefined.

File: src/requirements/requirement-frontend-utils.ts [hunk 4]
```diff
@@ -86,6 +89,37 @@ export function requirementAllowDoubleCounting(
...
+      // at least one requirement has the allowCourseDoubleCounting flag
+      allowCourseDoubleCounting ||
+      // or the source specific name is different
+      requirementA.sourceSpecificName !== requirementB.sourceSpecificName
+    );
+  }
+  // requirement source type is not the same
+  // exactly one source type is minor
+  if (requirementA.sourceType === 'Minor' || requirementB.sourceType === 'Minor') {
+    return true;
```
[the remark this answers is not in the document]
Comment (benjamin-shen) [reply src/requirements/requirement-frontend-utils.ts:111]: afaict, minors are specifically for undergrads.
image

I'm not sure how a major+minor+grad combination would work. I'm thinking you shouldn't be able to double count a course for a minor requirement and a grad requirement, but I could be wrong. @einc do you know?

File: src/requirements/requirement-frontend-utils.ts [hunk 5]
```diff
@@ -422,10 +452,12 @@ export function computeFulfillmentCoursesAndStatistics(
 
 export function getAllEligibleRelatedRequirementIds(
   courseId: number,
+  // uniqueId: number,
```
[the remark this answers is not in the document]
Comment (benjamin-shen) [reply src/requirements/requirement-frontend-utils.ts:455]: ok 😢

File: src/requirements/requirement-frontend-utils.ts [hunk 6]
```diff
@@ -439,6 +471,19 @@ export function getAllEligibleRelatedRequirementIds(
       }
       return [];
     });
+  // TODO @bshen fix this
```
[the remark this answers is not in the document]
Comment (benjamin-shen) [reply src/requirements/requirement-frontend-utils.ts:474]: yeah, i'll delete it in this PR

File: src/requirements/requirement-constraints-utils.ts [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+
+type RequirementFulfillmentGraphConstraintViolations<Requirement extends string> = {
+  constraintViolationsGraph: RequirementFulfillmentGraph<Requirement, CourseWithUniqueId>;
+  requirementsThatDoNotAllowDoubleCounting: Map<string | number, Set<Requirement[]>>;
+  doubleCountedCourseUniqueIDSet: Set<string | number>;
+};
+
+// utility functions for nested set computations
+const setEquals = <T>(a: Set<T>, b: Set<T>) =>
+  a.size === b.size && [...a].every(value => b.has(value));
```
[the remark this answers is not in the document]
Comment (benjamin-shen) [reply src/requirements/requirement-constraints-utils.ts:11]: [shameless](https://stackoverflow.com/a/44827922)

File: src/requirements/requirement-constraints-utils.ts [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+const setAdd = <T>(set: Set<T>, value: T) => {
```
[the remark this answers is not in the document]
Comment (benjamin-shen) [reply src/requirements/requirement-constraints-utils.ts:12]: I don't think it's necessary for now, but maybe we should if we find ourselves needing to refactor & use them in other places as well

File: src/requirements/requirement-frontend-computation.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (willespencer) [reply src/requirements/requirement-frontend-computation.ts:241]: Could you rename the former to something like `requirementsThatDoNotAllowDoubleCountingForCourses` to match the name of `requirementsThatDoNotAllowDoubleCountingForSingleCourse` ?

File: src/requirements/requirement-constraints-utils.ts [hunk 9]
```diff
@@ -0,0 +1,112 @@
+import RequirementFulfillmentGraph, { CourseWithUniqueId } from './requirement-graph';
+
+type RequirementFulfillmentGraphConstraintViolations<Requirement extends string> = {
+  constraintViolationsGraph: RequirementFulfillmentGraph<Requirement, CourseWithUniqueId>;
+  requirementsThatDoNotAllowDoubleCounting: Map<string | number, Set<Requirement[]>>;
```
[the remark this answers is not in the document]
Comment (benjamin-shen) [reply src/requirements/requirement-constraints-utils.ts:5]: It basically means the same thing. Where you see `string` is usually in places where the function is called / the value is returned (and generics aren't used).

File: src/requirements/requirement-frontend-utils.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (einc) [reply src/requirements/requirement-frontend-utils.ts:103]: [Yup](https://www.cornell.edu/academics/fields.cfm) just for UG. 
It seems like [graduate minors](https://gradschool.cornell.edu/academics/fields-of-study/field/sustainable-energy-minor/) also exist, but I don't think we need to worry about these for now.

**Review by willespencer — APPROVED** [review]
All changes LGTM!

**Pull request merged**